src/DesignTime/QuotationsFactory.fs

```diff
@@ -628,4 +629,68 @@ type internal QuotationsFactory private() =
 
             cmdProvidedType.AddMember resultSetsType
 
-
+    static member internal GetEnumType (enum: Enum) typeName =
+        let t = ProvidedTypeDefinition(typeName, Some typeof<string>, hideObjectMethods = true, nonNullable = true)
+        for value in enum.Values do t.AddMember(ProvidedField.Literal(value, t, value))
+        t
+
+    static member internal GetCompositeType (composite: Composite) typeName =
+        let dictType = typeof<IDictionary<string, obj>>
+        let containsKeyMi = dictType.GetMethod("ContainsKey", Reflection.BindingFlags.Instance ||| Reflection.BindingFlags.Public)
+        let itemMi = dictType.GetMethod("get_Item", Reflection.BindingFlags.Instance ||| Reflection.BindingFlags.Public)
+
+        let t = ProvidedTypeDefinition(typeName, Some typeof<obj>, hideObjectMethods = true, nonNullable = true)
+
+        let props, ctorParams =
+            composite.Fields
+            |> List.map (fun field ->
+                let clrType =
+                    if field.TypeName.StartsWith "_" then
+                        let elemType = getTypeMapping(field.TypeName.TrimStart('_') ) 
+                        elemType.MakeArrayType()
+                    else
+                        getTypeMapping(field.TypeName)
+
+                // composite type fields are always nullable
+                let clrTypeOption = typedefof<_ option>.MakeGenericType([| clrType |])
+
+                let someCase, noneCase = Utils.GetOptionCaseInfos clrTypeOption
```

Comment:
I've hit a bit of a wall here. Since fields on composite types are always nullable, provided properties need to return an option, which is created with reflection using `GetUnionCases`. The problem is that this method does not appear to work when the option-wrapped type is a provided type like an enum or a different composite, so I treat them as `obj`. @melanore, do you have any thoughts on this?